dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchMemoryRecordService.cs

```diff
@@ -0,0 +1,391 @@
+    /// </summary>
+    /// <param name="searchClient">The search client to use when uploading the document.</param>
+    /// <param name="collectionName">The name of the collection to upsert the records to.</param>
+    /// <param name="records">The records to upload.</param>
+    /// <param name="innerOptions">The azure ai search sdk options for uploading a document.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The document upload result.</returns>
+    private Task<Response<IndexDocumentsResult>> MapToStorageModelAndUploadDocumentAsync(
+        SearchClient searchClient,
+        string collectionName,
+        IEnumerable<TDataModel> records,
+        IndexDocumentsOptions innerOptions,
+        CancellationToken cancellationToken)
+    {
+        // Use the user provided mapper.
+        if (this._options.MapperType == AzureAISearchMemoryRecordMapperType.JsonObjectCustomMapper)
+        {
+            var jsonObjects = RunModelConversion(
+                () => records.Select(this._options.JsonObjectCustomMapper!.MapFromDataToStorageModel),
+                collectionName,
+                "UploadDocuments");
+
+            return RunOperationAsync(
+                () => searchClient.UploadDocumentsAsync<JsonObject>(jsonObjects, innerOptions, cancellationToken),
+                collectionName,
+                "UploadDocuments");
+        }
+
+        // Use the built in Azure AI Search mapper.
+        return RunOperationAsync(
+            () => searchClient.UploadDocumentsAsync<TDataModel>(records, innerOptions, cancellationToken),
+            collectionName,
+            "UploadDocuments");
+    }
+
+    /// <summary>
+    /// Get a search client for the index specified.
+    /// Note: the index might not exist, but we avoid checking everytime and the extra latency.
+    /// </summary>
+    /// <param name="indexName">Index name</param>
+    /// <returns>Search client ready to read/write</returns>
+    private SearchClient GetSearchClient(string indexName)
+    {
+        // Check the local cache first, if not found create a new one.
+        if (!this._searchClientsByIndex.TryGetValue(indexName, out SearchClient? client))
+        {
+            client = this._searchIndexClient.GetSearchClient(indexName);
+            this._searchClientsByIndex[indexName] = client;
+        }
+
+        return client;
+    }
+
+    /// <summary>
+    /// Choose the right collection name to use for the operation by using the one provided
+    /// as part of the operation options, or the default one provided at construction time.
+    /// </summary>
+    /// <param name="operationCollectionName">The collection name provided on the operation options.</param>
+    /// <returns>The collection name to use.</returns>
+    private string ChooseCollectionName(string? operationCollectionName)
```

Comment:
nit: `GetCollectionName`?

Reply:
Get makes it sound like I'm only retrieving it, but here I'm actually making a choice, so I think choose is more descriptive.